tijanabekic/assignment3/Assignment 3,new.java

```diff
@@ -0,0 +1,167 @@
+//simple dictionary for test purpose, not the follow up solution
+public class Dictionary {
+
+    private static LinkedList<String> words;
+    private static LinkedList<String> prefixes;
+
+    public Dictionary() {
+        prefixes = new LinkedList<>();
+        prefixes.add("CARD");
+        prefixes.add("CART");
+        prefixes.add("CAR");
+        prefixes.add("CAT");
+        prefixes.add("CA");
+        prefixes.add("C");
+
+        words = new LinkedList<>();
+        words.add("CARD");
+        words.add("CART");
+        words.add("CAR");
+        words.add("CAT");
+    }
+
+    public static boolean isPrefix(String word) {
+        for (String s : prefixes) {
+            if (word.equals(s)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public static boolean isWord(String word) {
+
+        for (String s : words) {
+            if (word.equals(s)) {
+                return true;
+            }
+        }
+        return false;
+    }
+    
+    public void addWord(String word) {
+        words.add(word);
+        
+    }
+
+}
+
+
+public class A3Q1 {
+    public static Dictionary dictionary;
+    public static void checkWord(char[][] grid, int i, int j, String start, int m, int n, HashMap<String, Boolean> map, HashSet<String> set) {
+        
+        //check if out of bounds
+        if (i < 0 || j < 0 || i >= m || j >= n) {
+            return;
+        }
+        //make the string for checking
+        start =  start + grid[i][j];
+        
+        //add to map if the word is not there
+        if (!map.containsKey(start)) {

```

Comment:
What are you using map for here? You only seem to check it immediately afterwards so could just check for `if (!dictionary.isPrefix(start)) { return; }`
It changes the O(log dictionaryLength) into an O(1) operation after the second call I guess.